File: win/directx/sdk-9.0/Bin/DXUtils/Visual Studio 7.0 Wizards/VB .NET/VBWizards/Mdx9AppWizard/Templates/1033/dplay.vb
Imports System
Imports System.Collections
Imports System.Windows.Forms
Imports Microsoft.DirectX
Imports Microsoft.DirectX.DirectPlay
[!if USE_VOICE]
Imports Voice = Microsoft.DirectX.DirectPlay.Voice
[!endif]

'/ <summary>
'/ The Class that houses DirectPlay code for the application.
'/ </summary>
Public Class PlayClass
    Private message As MessageDelegate = Nothing

[!if GRAPHICSTYPE_DIRECT3D]
    Private parent As GraphicsClass = Nothing
[!else]
    Private parent As MainClass = Nothing
[!endif]

    Public  peerObject As Peer = Nothing
[!if USE_VOICE]
    Private voiceClient As Voice.Client = Nothing
    Private voiceServer As Voice.Server = Nothing
    Private voiceWizard As VoiceWizard = Nothing
[!endif]
    Private Connect As ConnectWizard = Nothing
    Private PlayerList As ArrayList = New ArrayList()
    Private LocalPlayerID As Integer = 0
[!if GRAPHICSTYPE_DIRECT3D]
[!if USE_VOICE]
    Public AppGuid As Guid = New Guid("876a3036-ffd7-46bc-9209-b42f617b9bF0")
[!else]
    Public AppGuid As Guid = New Guid("{876a3036-ffd7-46bc-9209-b42f617b9bF1}")
[!endif]
[!endif]
[!if GRAPHICSTYPE_DIRECTDRAW || GRAPHICSTYPE_NOGRAPHICS]
[!if USE_VOICE]
    Public AppGuid As Guid = New Guid("{876a3036-ffd7-46bc-9209-b42f617b9bF2}")
[!else]
    Public AppGuid As Guid = New Guid("{876a3036-ffd7-46bc-9209-b42f617b9bF3}")
[!endif]
[!endif]

    Public Structure Players
        Public dpnID As Integer 
        Public Name As string 
        Public Sub New(id As Integer , n As string)
            dpnID = id 
            Name = n
        End Sub
    End Structure

[!if GRAPHICSTYPE_DIRECT3D]
    Public Sub New(parent As GraphicsClass)
[!else]
    Public Sub New(parent As MainClass)
[!endif]
        Me.parent = parent
        Me.peerObject = peerObject
        Me.message = New MessageDelegate(AddressOf parent.MessageArrived)

        peerObject = New Peer()
        ' First set up our event handlers (We only need events for the ones we care about)
        AddHandler peerObject.PlayerCreated, AddressOf Me.PlayerCreated
        AddHandler peerObject.PlayerDestroyed, AddressOf  Me.PlayerDestroyed
        AddHandler peerObject.HostMigrated, AddressOf Me.HostMigrated
        AddHandler peerObject.Receive, AddressOf Me.DataReceived
        AddHandler peerObject.SessionTerminated, AddressOf Me.SessionTerminated

        Connect = New ConnectWizard(peerObject, AppGuid, "[!output PROJECT_NAME]")
        If (Not Connect.StartWizard()) Then
            Me.parent.Show()
            MessageBox.Show("DirectPlay initialization was incomplete. Application will terminate.")
            Throw New DirectXException()
        End If
[!if USE_VOICE]
        voiceClient = New Voice.Client(peerObject)
        voiceWizard = New VoiceWizard()

        If (Connect.IsHost) Then
            'create the voice server
            voiceServer = New Voice.Server(peerObject)

            'init the voice server
            voiceWizard.InitVoiceHost(peerObject, voiceServer, voiceClient, parent)
        Else
            ' Connect as a client.
            voiceWizard.InitVoiceClient(peerObject, voiceClient, parent)
        End If
[!endif]
    End Sub

    Public Sub WriteMessage(msg As Byte)
        Dim packet As NetworkPacket = New NetworkPacket()

        packet.Write(msg)
        peerObject.SendTo(PlayerID.AllPlayers, packet, 0, SendFlags.Guaranteed)
    End Sub

    #Region "DirectPlayEvents"
    '/ <summary>
    ' These are the events the app will handle
    ' when DPlay fires them.
    '/ </summary>
    Private Sub PlayerCreated(sender As Object, dpMessage As PlayerCreatedEventArgs)
        ' Get the PlayerInformation and store it 
        Dim dpPeer As PlayerInformation = peerObject.GetPeerInformation(dpMessage.Message.PlayerID)
        Dim oPlayer As Players = New Players(dpMessage.Message.PlayerID,dpPeer.Name)
        ' We lock the data here since it is shared across multiple threads.
        SyncLock(PlayerList)
            PlayerList.Add(oPlayer)
        End SyncLock
        ' Save Me player id if it's ourselves
        If (dpPeer.Local)
            LocalPlayerID = dpMessage.Message.PlayerID
        End If
    End Sub
    Private Sub PlayerDestroyed(sender As Object, dpMessage As PlayerDestroyedEventArgs)
        ' Remove Me player from our list
        ' We lock the data here since it is shared across multiple threads.
        SyncLock (PlayerList)
            Dim player As Players 
            For Each player In PlayerList
                If (dpMessage.Message.PlayerID = player.dpnID)
                    PlayerList.Remove(player)
                    Exit For
                End If
            Next
        End SyncLock
    End Sub
    Private Sub HostMigrated(sender As Object, dpMessage As HostMigratedEventArgs)
        If (LocalPlayerID = dpMessage.Message.NewHostID) 
            ' This application instance is the New host, update the UI
            parent.Text += " (HOST)"
        End If
    End Sub
    Private Sub DataReceived(sender As Object, dpMessage As ReceiveEventArgs)
        SyncLock(Me)
            Dim data as Byte = dpMessage.Message.ReceiveData.Read(GetType(Byte))
            message(data)
        End SyncLock
    End Sub
    Private Sub SessionTerminated(sender As Object, dpMessage As SessionTerminatedEventArgs)
        ' This session is being terminated, let the user know
        If (dpMessage.Message.ResultCode = ResultCode.HostTerminatedSession)
            MessageBox.Show("The Host has terminated Me session.  This sample will now exit.", "Exiting", MessageBoxButtons.OK, MessageBoxIcon.Information)
        Else
            MessageBox.Show("The session has been lost.  This sample will now exit.", "Exiting", MessageBoxButtons.OK, MessageBoxIcon.Information)
        End If

        ' This will post a message on the main thread to shut down our form
        parent.BeginInvoke(New PeerCloseCallback(AddressOf parent.PeerClose))
    End Sub
    #End Region
End Class

File: win/directx/sdk-9.0/Bin/DXUtils/Visual Studio 7.0 Wizards/C#/VC#Wizards/Mdx9AppWizard/Templates/1033/dplay.cs
using System;
using System.Collections;
using System.Windows.Forms;
using Microsoft.DirectX;
using Microsoft.DirectX.DirectPlay;
[!if USE_VOICE]
using Voice = Microsoft.DirectX.DirectPlay.Voice;
[!endif]

/// <summary>
/// The class that houses DirectPlay code for the application.
/// </summary>
public class PlayClass
{
    private MessageDelegate message = null;

[!if GRAPHICSTYPE_DIRECT3D]
    private GraphicsClass parent = null;
[!else]
    private MainClass parent = null;
[!endif]

    public  Peer peerObject = null;
[!if USE_VOICE]
    private Voice.Client voiceClient = null;
    private Voice.Server voiceServer = null;
    private VoiceWizard voiceWizard = null;
[!endif]
    private ConnectWizard Connect = null;
    private ArrayList PlayerList = new ArrayList();
    private int LocalPlayerID = 0;
[!if GRAPHICSTYPE_DIRECT3D]
[!if USE_VOICE]
    public Guid AppGuid = new Guid(0x876a3036, 0xffd7, 0x46bc, 0x92, 0x9, 0xb4, 0x2f, 0x61, 0x7b, 0x9b, 0xF0);
[!else]
    public Guid AppGuid = new Guid(0x876a3036, 0xffd7, 0x46bc, 0x92, 0x9, 0xb4, 0x2f, 0x61, 0x7b, 0x9b, 0xF1);
[!endif]
[!endif]
[!if GRAPHICSTYPE_DIRECTDRAW || GRAPHICSTYPE_NOGRAPHICS]
[!if USE_VOICE]
    public Guid AppGuid = new Guid(0x876a3036, 0xffd7, 0x46bc, 0x92, 0x9, 0xb4, 0x2f, 0x61, 0x7b, 0x9b, 0xF2);
[!else]
    public Guid AppGuid = new Guid(0x876a3036, 0xffd7, 0x46bc, 0x92, 0x9, 0xb4, 0x2f, 0x61, 0x7b, 0x9b, 0xF3);
[!endif]
[!endif]

    public struct Players
    {
        public int dpnID;
        public string Name;
        public Players(int id, string n)
        { dpnID = id; Name = n; }
    }       

[!if GRAPHICSTYPE_DIRECT3D]
    public PlayClass(GraphicsClass parent)
[!else]
    public PlayClass(MainClass parent)
[!endif]
    {
        this.parent = parent;
        this.peerObject = peerObject;
        this.message = new MessageDelegate(parent.MessageArrived);

        peerObject = new Peer();
        // First set up our event handlers (We only need events for the ones we care about)
        peerObject.PlayerCreated += new PlayerCreatedEventHandler(this.PlayerCreated);
        peerObject.PlayerDestroyed += new PlayerDestroyedEventHandler(this.PlayerDestroyed);
        peerObject.HostMigrated += new HostMigratedEventHandler(this.HostMigrated);
        peerObject.Receive += new ReceiveEventHandler(this.DataReceived);
        peerObject.SessionTerminated += new SessionTerminatedEventHandler(this.SessionTerminated);

        Connect = new ConnectWizard(peerObject, AppGuid, "[!output PROJECT_NAME]");
        if (!Connect.StartWizard())
        {
            MessageBox.Show("DirectPlay initialization was incomplete. Application will terminate.");
            throw new DirectXException();
        }
[!if USE_VOICE]
        voiceClient = new Voice.Client(peerObject); 
        voiceWizard = new VoiceWizard();

        if (Connect.IsHost)
        {
            //create the voice server
            voiceServer = new Voice.Server(peerObject);

            //init the voice server
            voiceWizard.InitVoiceHost(peerObject, voiceServer, voiceClient, parent);
        }
        else
        {
            // Connect as a client.
            voiceWizard.InitVoiceClient(peerObject, voiceClient, parent);
        }
[!endif]
    }

    public void WriteMessage(byte msg)
    {
        NetworkPacket packet = new NetworkPacket();

        packet.Write(msg);
        peerObject.SendTo((int)PlayerID.AllPlayers, packet, 0, SendFlags.Guaranteed);
    }

    #region DirectPlayEvents
    /// <summary>
    // These are the events the app will handle
    // when DPlay fires them.
    /// </summary>
    private void PlayerCreated(object sender, PlayerCreatedEventArgs dpMessage)
    {
        // Get the PlayerInformation and store it 
        PlayerInformation dpPeer = peerObject.GetPeerInformation(dpMessage.Message.PlayerID);
        Players oPlayer = new Players(dpMessage.Message.PlayerID,dpPeer.Name);
        // We lock the data here since it is shared across multiple threads.
        lock (PlayerList)
        {
            PlayerList.Add(oPlayer);
        }
        // Save this player id if it's ourselves
        if (dpPeer.Local)
            LocalPlayerID = dpMessage.Message.PlayerID;

    }
    private void PlayerDestroyed(object sender, PlayerDestroyedEventArgs dpMessage)
    {
        // Remove this player from our list
        // We lock the data here since it is shared across multiple threads.
        lock (PlayerList)
        {
            foreach (Players player in PlayerList)
            {
                if (dpMessage.Message.PlayerID == player.dpnID)
                {
                    PlayerList.Remove(player);
                    break;
                }
            }
        }
    }
    private void HostMigrated(object sender, HostMigratedEventArgs dpMessage)
    {
        if (LocalPlayerID == dpMessage.Message.NewHostID) 
        {
            // This application instance is the new host, update the UI
            parent.Text += " (HOST)";
        }
    }
    private void DataReceived(object sender, ReceiveEventArgs dpMessage)
    {
        lock(this)
        {
            byte data = (byte)dpMessage.Message.ReceiveData.Read(typeof(byte));
            message(data);
        }
    }
    private void SessionTerminated(object sender, SessionTerminatedEventArgs dpMessage)
    {
        // This session is being terminated, let the user know
        if (dpMessage.Message.ResultCode == ResultCode.HostTerminatedSession)
            MessageBox.Show("The Host has terminated this session.  This sample will now exit.", "Exiting", MessageBoxButtons.OK, MessageBoxIcon.Information);
        else
            MessageBox.Show("The session has been lost.  This sample will now exit.", "Exiting", MessageBoxButtons.OK, MessageBoxIcon.Information);

        // This will post a message on the main thread to shut down our form
        parent.BeginInvoke(new PeerCloseCallback(parent.PeerClose));
    }
    #endregion
}